arch/lkl/kernel/signal.c

```diff
@@ -28,47 +31,249 @@ static void initialize_uctx(struct ucontext *uctx, const struct pt_regs *regs)
+    }
+
+    next = node->next;
+    task->signal_list_head = next;      /* drop the head */
+    if (next == NULL)                   /* was that the last node? */
+        task->signal_list_tail = NULL;  /* if so there is now no tail */
+        
+    
+    spin_unlock(&task->signal_list_lock);
+    
+    memcpy(sig, &node->sig, sizeof(*sig)); /* copy the signal back to the caller */
+    LKL_TRACE("Fetching task %p signal %d\n", current, sig->sig);
+    kfree(node);
+    return 1;
+}    
+
+
+/*
+ *  This is the sync signal case and that signal will have just been forced into the
+ *  regular task queue of signals so the pre-existing code will work. See lkl_do_trap
+ *
+ *  Up for debate though is whether it would be better to either directly invoke
+ *  the handler in lkl_do_trap or here to have versions of move_signals_to_task, send_current_signals
+ *  that operate by scanning the list for signr, the given signal.
+ */
+
+/* This is always called while owning the cpu */
+
 void lkl_process_trap(int signr, struct ucontext *uctx)
 {
     struct ksignal ksig;
 
+    /* In DEBUG build check that the lock is currently held by this task */
+    lkl_assert_cpu_owner();
+
     LKL_TRACE("enter\n");
 
-    while (get_signal(&ksig)) {
-        LKL_TRACE("ksig.sig=%d", ksig.sig);
+    /*
+     *  Capture and queue pending signals and traps.
+     *  Expect at least the one 'signr'.
+     */
 
+    move_signals_to_task();
+    
+    /*
+     *  Walk that queue and find the first one of the type we expect.
+     *  Note, no expectation that a trap will happen in order wrt to any
+     *  async signal already queued.
+     */
+
+    while (get_given_ksignal(task_thread_info(current), &ksig, signr)) {
         /* Handle required signal */
-        if(signr == ksig.sig)
+        if (signr == ksig.sig)
         {
+            LKL_TRACE("trap task %p %d\n", current, signr);
             handle_signal(&ksig, uctx);
             break;
         }
     }
 }
 
-void do_signal(struct pt_regs *regs)
+/*
+ * Get the end of the list and make it point at the new node,
+ * if there were no nodes before then bot head and tail point at this one node
+ */
+
+void append_ksignal_node(struct thread_info *task, struct ksignal *ksig)
+{
+    struct ksignal_list_node *ptr;
+    
+    /* first make a place to hold our signal */
+    struct ksignal_list_node *node = kmalloc(sizeof(struct ksignal_list_node), GFP_ATOMIC);     
+    if (node == NULL) {
+        lkl_bug("Unable to allocate memory to hold signal state.");
+    }
+    memcpy(&node->sig, ksig, sizeof(*ksig));
+    node->next = NULL;
+
+    /* now put it in the linked list */
+    spin_lock(&task->signal_list_lock);
+    ptr = task->signal_list_tail;
+    if (ptr == NULL) { /* list is empty so point both head and tail at the given node */
+        task->signal_list_head = node;
+        task->signal_list_tail = node;
+    } else { /* otherwise just the tail changes */
+        ptr->next = node; 
```

Comment:
Does the tail need to be updated here? For example:

```
task->signal_list_tail = node;
```

Reply:
Good catch, well spotted!

Reply:
Fixed now.

Reply:
Thank you.